File: 409. Longest Palindrome.cpp
class Solution {
public:
    // hashtable: O(N)time, O(1)space, 32.77%
    int longestPalindrome(string s) {
        if(s.empty()) return 0;
        int ret = 0;
        bool odd = false;
        unordered_map<char, int> letter;
        for(auto c: s) letter[c]++;
        for(auto it = letter.begin(); it != letter.end(); it++){
            if(it->second % 2){
                odd = true;
                ret += it->second - 1;  // bug: miss this line
            }
            else ret += it->second;
        }
        return odd? ret + 1 : ret;
    }
};

File: 409. Longest Palindrome.py
class Solution(object):
    
    # hash table: O(N) time & space
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: int
        """
        count, ret = {}, 0
        for c in s:
            count[c] = count.get(c, 0) + 1
        hasOdd = False
        for cnt in count.values():
            if cnt % 2 != 0:
                hasOdd = True
            ret += cnt / 2 * 2  # odd can contribute E.g: 3 can contribute 2
        return ret + (1 if hasOdd else 0)
        